### PR TITLE
Precondition that read doesn't return EINVAL

### PR DESCRIPTION
**Pull request by FranzBusch**
### Motivation:

`read` should never return `EINVAL`. If it does return it though it is a bug inside NIO and we should precondition for it similar to what we do for `EBADF` & `EFAULT`. Closes https://github.com/apple/swift-nio/issues/1339

### Modifications:

This PR adds a new precondition for `EINVAL` and adds test that exercise this precondition.

### Result:

We are now preconditioning against `read` returning `EINVAL`.

### CODE REVIEW

File: IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh [hunk 1]
```diff
@@ -61,7 +63,7 @@ EOF
 make_package
 
 for mode in debug release; do
-    for error in EFAULT EBADF; do
+    for error in EFAULT EBADF, EINVAL; do
```
Comment (weissi) [inline IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh:66]: The comma isn’t correct here

Comment (FranzBusch) [inline IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh:66]:
```suggestion
    for error in EFAULT EBADF EINVAL; do
```

**Comment by FranzBusch** [comment]
@swift-nio-bot test please

**Comment by FranzBusch** [comment]
@swift-server-bot test this please

**Comment by FranzBusch** [comment]
@weissi Do you have any idea why these tests are failing? They are also failing for me locally on main and I cannot figure out why they are returning the wrong value now.

File: IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh [hunk 2]
```diff
@@ -47,10 +47,12 @@ if pipeErr != 0 {
 }
 let makeEBADFHappen =  CommandLine.arguments.dropFirst().first == .some("EBADF")
 let makeEFAULTHappen = CommandLine.arguments.dropFirst().first == .some("EFAULT")
-var whatevs: UInt8 = 123
+let makeEINVALHappen = CommandLine.arguments.dropFirst().first == .some("EINVAL")
+var whatevs: UInt8 = makeEINVALHappen ? 3021 : 123
```
Comment (Lukasa) [inline IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh:51]: This does not produce EINVAL on Posix.write. From the manual page:

```
[EINVAL]           The pointer associated with fildes is negative.
```

`3021` is not negative.

Comment (Lukasa) [reply IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh:51]: It is, however, larger than `UInt8` can hold, so the compile fails. This is why the tests are failing.

File: IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh [hunk 3]
```diff
@@ -47,10 +47,12 @@ if pipeErr != 0 {
 }
 let makeEBADFHappen =  CommandLine.arguments.dropFirst().first == .some("EBADF")
 let makeEFAULTHappen = CommandLine.arguments.dropFirst().first == .some("EFAULT")
-var whatevs: UInt8 = 123
+let makeEINVALHappen = CommandLine.arguments.dropFirst().first == .some("EINVAL")
+var whatevs: UInt8 = makeEINVALHappen ? -1 : 123
```
Comment (Lukasa) [inline IntegrationTests/tests_02_syscall_wrappers/test_02_unacceptable_errnos.sh:51]: This still won't work. `EINVAL` is triggered by the _pointer_ being negative: this is the `pointee`, and so unaffected. Also, this still won't compile. :smile:

File: Tests/NIOPosixTests/EventLoopTest+XCTest.swift [hunk 4]
```diff
@@ -2,7 +2,7 @@
 //
 // This source file is part of the SwiftNIO open source project
 //
-// Copyright (c) 2017-2021 Apple Inc. and the SwiftNIO project authors
+// Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
```
Comment (FranzBusch) [inline Tests/NIOPosixTests/EventLoopTest+XCTest.swift:5]: This is currently breaking on the CI and was caused by my previous PR where the CI did not detect this necessary change. I am just going to include it here to fix the CI.

**Review by Lukasa — CHANGES_REQUESTED** [review]
This change is good, but it's a bit too narrow and a bit too broad.

The original issue only asked to forbid `EINVAL` on `read`. This change adds it to all syscalls in NIOCore, but does not affect those in NIOPosix, which is where the relevant `read` is actually made. As a result, it does not resolve the original issue.

Additionally, the change as written affects _all_ system calls. However, some EINVALs we actually want to tolerate. As an example, see #1030. As a result, we likely need a more complex solution that only adds EINVAL policing to some specific system calls, not all of them.

File: IntegrationTests/tests_02_syscall_wrappers/test_03_unacceptable_read_errnos.sh [hunk 5]
```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+##===----------------------------------------------------------------------===##
+##
+## This source file is part of the SwiftNIO open source project
+##
+## Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
```
Comment (Lukasa) [inline IntegrationTests/tests_02_syscall_wrappers/test_03_unacceptable_read_errnos.sh:6]:
```suggestion
## Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
```

File: Sources/NIOPosix/System.swift [hunk 6]
```diff
@@ -411,14 +386,14 @@ internal enum Posix {
 
     @inline(never)
     internal static func read(descriptor: CInt, pointer: UnsafeMutableRawPointer, size: size_t) throws -> IOResult<ssize_t> {
-        return try syscall(blocking: true) {
+        return try syscall(blocking: true, unacceptableErrnoCodes: defaultUnacceptableErrnoCodes + [EINVAL]) {
```
Comment (Lukasa) [inline Sources/NIOPosix/System.swift:389]: Unfortunately I don't think this approach is sufficient for us.

The biggest issue I have here is that this causes a runtime Array allocation on all codepaths through this function. I suspect the performance impact of this is fairly substantial.

Unfortunately, the only really effective solution at the moment is to duplicate the `isUnacceptableErrno` function with a new body. This is what we did for `close`, and sadly is what we have to do here as well.

**Review by Lukasa — CHANGES_REQUESTED** [review]

File: Sources/NIOPosix/System.swift (hunk 6 above)
Comment (Lukasa) [reply Sources/NIOPosix/System.swift:389]: Ah, yup, in fact the allocation counter tests failed.

**Comment by Lukasa** [comment]
@swift-nio-bot test perf please

**Comment by Lukasa** [comment]
Yup, that report above is not right: those are clearly the results from @weissi's previous PR.

**Comment by Lukasa** [comment]
@tomerd Seems like the NIO performance builds aren't working again.

File: Sources/NIOPosix/System.swift [hunk 7]
```diff
@@ -195,19 +220,32 @@ internal func syscall<T: FixedWidthInteger>(blocking: Bool,
...
+
+/*
+ * Sorry, we really try hard to not use underscored attributes. In this case
+ * however we seem to break the inlining threshold which makes a system call
+ * take twice the time, ie. we need this exception.
+ */
 @inline(__always)
-internal func wrapErrorIsNullReturnCall<T>(where function: String = #function, _ body: () throws -> T?) throws -> T {
+@discardableResult
+internal func sysReadCall<T: FixedWidthInteger>(where function: String = #function,
```
Comment (Lukasa) [inline Sources/NIOPosix/System.swift:231]: Hmm, I don't love this: it's too specific to `read` (same with those above). Maybe `syscallForbiddingEINVAL`?

**Review by Lukasa — APPROVED** [review]
LGTM, nice one!

**Pull request merged**